Catappult/appcoins-wallet-android
# [APPC-2231] improve offline first

**What does this PR do?**

   Offline first logic - to use local content (DB, caches) and display that before having the more recent content, retrieved from API - for Promotions and Gamification screen.
   This implementation follows what was done for subscriptions (see [here](https://github.com/Catappult/appcoins-wallet-android/blob/feature-branch/Subscriptions/app/src/main/java/com/asfoundation/wallet/subscriptions/UserSubscriptionRepository.kt))
   Some isolated contexts that used promotion-related information (e.g. wallet origin, user level) were "updated" with this new logic, even though they may not actually use DB content.
   Finally, this also featured changes regarding currency conversion rates. The rates from appc/eth to fiat currency are now being saved in a local Database, each time the api for conversion is called. Some extra measures were taken to make sure rates are not always invalidated when one wants to convert 0 appc/eth to fiat. See comment on method `saveRate` of `RoomCurrencyConversionRatesPersistence.kt` for more information.

**Database changed?**

   No

**Where should the reviewer start?**

You can start by checking repository changes and unit tests. First files to see, in this order:
- [ ] BdsPromotionsRepository.kt
- [ ] Gamification.kt
- [ ] GamificationTest.kt

**How should this be manually tested?**

  You can start by running the unit tests to understand the basics of this logic.
  Test the isolated contexts (register user on intercom, rakam superproperties, etc.) that were "adapted" to the new logic. See if they still work as intented.
  Even though other flows that make conversion calls to the api and don't use cached conversion rates shouldn't be affected, they should be tested all the same, to make sure they still function the same way as before. These places are: Home and My Wallets screen (to convert balances in appc/eth); topup and in app purchase screens - payment method selection and adyen payment (to convert APPC values to fiat).
  Finally, test the promotions and gamification screen, testing with and without network, and for different kinds of users (unknown, aptoide). If you feel like you can't tell a difference between DB and API, this might be because DB already has the current info (being synced with API in transactions screen or perks notification service), or because API response is too fast (responde may be cached server-side, from previous requests).

**What are the relevant tickets?**

  Tickets related to this pull-request: [APPC-2231](https://aptoide.atlassian.net/browse/APPC-2231)
  Unlike originally specified in the ticket, this PR will not feature offline first in transactions screen, that will be done in [APPC-2330](https://aptoide.atlassian.net/browse/APPC-2330)

**Questions:**

   Does this add new dependencies which need to be added? (Eg. new libs, new keys, etc.)
   No


**Code Review Checklist**
